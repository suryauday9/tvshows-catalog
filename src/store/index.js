import Vue from "vue";
import Vuex from "vuex";

import {
getAllShows,
getShowDetails,
getShowThumbnails,
getShowsBySearch
} from "../services/service";

Vue.use(Vuex);

const state = {
  shows: [],
  genres: [],
  showDetails: [],
  showThumbnails: [],
  showSearchbar: false
};

export const sortByRating = (response) => {
  return response.sort((prevValue, nextValue) =>
  prevValue.rating.average < nextValue.rating.average ? 1 : -1);
};

export const getGenres = (response) => {
  return response
  .reduce((result, show) => result.concat(show.genres), [])
  .filter((genre, index, self) => self.indexOf(genre) === index);
};

export const actions = {
  async getShowslist({ commit }) {
    getAllShows().then((response) => {
      commit('setShowslist', sortByRating(response));
      commit('setGenreslist', getGenres(response));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting all shows: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowDetails({ commit }, showId) {
    getShowDetails(showId).then((response) => {
      commit('setShowDetails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show details: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowThumbnails({ commit }, showId) {
    await getShowThumbnails(showId).then((response) => {
      commit('setShowThumbnails', response);
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem getting show thumbnails: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  async getShowsBySearch({ commit }, searchTerm) {
    await getShowsBySearch(searchTerm).then((response) => {
      let resultSet = response.map(result => (result.show));
      commit('setShowslist', resultSet);
      commit('setGenreslist', getGenres(resultSet));
    })
    .catch((error) => {
      const errorNotification = {
        type: 'error',
        message: `There was a problem in show search: ${error.message}`,
      };
      console.log(errorNotification);
    });
  },
  getSearchbarStatus ({ commit }) {
    commit('setSearchbarStatus', !this.state.showSearchbar);
  }
};

export const mutations = {
  setShowslist(state, payload) {
    state.shows = payload;
  },
  setGenreslist(state, payload) {
      state.genres = payload;
  },
  setShowDetails(state, payload) {
    state.showDetails = payload;
  },
  setShowThumbnails(state, payload) {
    state.showThumbnails = payload;
  },
  setSearchbarStatus(state, payload) {
    state.showSearchbar = payload;
  }
};

export default new Vuex.Store({
  state: state,
  actions: actions,
  mutations: mutations,
  modules: {}
});
