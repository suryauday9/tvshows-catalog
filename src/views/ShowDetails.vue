<template>
  <div class="show-details">
    <v-btn class="back-button black white--text" icon @click="goBack()">
      <v-icon>mdi-keyboard-return</v-icon>
    </v-btn>
    <div class="ma-5" v-bind:style="{ backgroundSize: '100% 96%', backgroundImage: 'url(' + backgroundImage + ')' }">
      <v-container>
          <v-row class="black--text py-md-12">
            <v-col class="show-thumbnail-block" cols="12" md="5">
              <v-img
                :src="posterImage"
                class="ml-auto elevation-6"
                contain
                width="390"
              ></v-img>
            </v-col>
            <v-col cols="12" md="7" class="pa-4 yellow lighten-5">
              <h1 class="display-2">
                {{ showDetails.name }}
              </h1>
              <div class="d-flex flex-column pt-2">
                <p class="subtitle-1 ma-0">
                    <v-chip class="white--text purple ma-2">
                      <v-icon small left>mdi-information</v-icon>
                      {{ showDetails.language }}
                    </v-chip>
                </p>    
                <p class="subtitle-1">
                  <v-chip class="white--text teal ma-2" v-for="genre in showDetails.genres" :key="genre">
                      <v-icon small left>mdi-information</v-icon>
                      {{ genre }}
                  </v-chip>
                </p>
                <p
                  class="headline d-flex align-center"
                  v-if="showDetails.rating && showDetails.rating.average"
                >
                  {{ showDetails.rating.average }}
                  <v-icon color="red">mdi-star</v-icon>
                </p>
                <div class="text-left" v-html="showDetails.summary"></div>
              </div>
            </v-col>
          </v-row>
        </v-container>
    </div>
  </div>
</template>

<script>
export default {
  name: "ShowDetails",
  components: {},
  data() {
    return {
      id: this.$route.params.id
    };
  },
  computed: {
    showDetails() {
        return this.$store.state.showDetails;
    },
    showThumbnails() {
        return this.$store.state.showThumbnails;
    },
    backgroundImage() {
      if (this.showThumbnails.length) {
        const bgImage = this.showThumbnails.filter(
          image => image.type === 'background',
        )[0];
        return bgImage ? bgImage.resolutions.original.url : '';
      }
      return '';
    },
    posterImage() {
      if (this.showThumbnails.length) {
        const psImage = this.showThumbnails.filter(
          image => image.type === 'poster',
        )[0];
        return psImage ? psImage.resolutions.original.url : '';
      }
      return '';
    },
  },
  methods: {
    initialize() {
        this.$store.dispatch('getShowDetails', this.id);
        this.$store.dispatch('getShowThumbnails', this.id);
    },
    goBack() {
      this.$router.back();
    }
  },
  beforeCreate() {
    if(!this.$route.params.id) {
      this.$router.push({ name: "showsCatalog"});
    }
  },
  created() {
    this.initialize();
  }
};
</script>
<style scoped>
.back-button {
  display: block;
  margin: 5px 16px;
}
.show-thumbnail-block {
  padding: 0 12px;
}
</style>
