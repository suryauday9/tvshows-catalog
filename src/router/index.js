import Vue from "vue";
import VueRouter from "vue-router";

Vue.use(VueRouter);

const routes = [
  {
    path: "/",
    name: 'ShowsCatalog',
    component: () => import('../views/ShowsCatalog.vue')
  },
  {
    path: '/show',
    name: 'show',
    component: () =>
      import('../views/ShowDetails.vue'),
    props: true,
  },
  {
    path: '/404',
    alias: '*',
    name: 'notFound',
    component: () =>
      import('../views/Page404.vue'),
  },
];

const router = new VueRouter({
  mode: "history",
  base: process.env.BASE_URL,
  routes
});

export default router;
