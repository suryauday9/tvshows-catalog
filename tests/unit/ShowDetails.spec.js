import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Vuetify from 'vuetify';
import VueRouter from 'vue-router';

import { initializeApp } from './initializeApp';
import { showsData, showDetails, showThumbnails } from './response';
import ShowDetails from '../../src/views/ShowDetails.vue';
import { routes } from '../../src/router';

const vuetify = new Vuetify();
const localVue = initializeApp(createLocalVue());

describe('ShowDetails component', () => {  
    let wrapper;
    const router = new VueRouter(routes);
    
    const store = new Vuex.Store({
        state: {
            shows: showsData,
            genres: [],
            showDetails: showDetails,
            showThumbnails: showThumbnails,
            showSearchbar: false
        },
        actions: {
            getShowDetails: jest.fn(),
            getShowThumbnails: jest.fn()
        }
    });

    const mountComponent = (options) => {
        return shallowMount(ShowDetails, {
            localVue,
            router,
            store,
            vuetify,
            ...options,
        })
    }

    beforeEach(() => {
        wrapper = mountComponent();
    })

    afterEach(() => {
        wrapper.destroy();
    })

    it('renders component correctly', () => {
        // see if component renders
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should be a Vue instance', () => {
        expect(wrapper.isVueInstance).toBeTruthy();
    });

    it('rendered component data should match with store', () => {
        expect(wrapper.vm.showDetails).toBe(store.state.showDetails);
        expect(wrapper.vm.showThumbnails).toBe(store.state.showThumbnails);
    });

    it('background image data should match with store', () => {
        const bgImg = store.state.showThumbnails.filter(
            image => image.type === 'background',
        );
        expect(wrapper.vm.backgroundImage).toBe(bgImg[0].resolutions.original.url);
    });

    it('background image & poster image should match with store values', () => {
        const store = new Vuex.Store({
            state: {
                shows: showsData,
                genres: [],
                showDetails: showDetails,
                showThumbnails: [],
                showSearchbar: false
            },
            actions: {
                getShowDetails: jest.fn(),
                getShowThumbnails: jest.fn()
            }
        });
        const dummyWrapper = shallowMount(ShowDetails, {
            localVue,
            router,
            store,
            vuetify
        });
        let bgImg = (store.state.showThumbnails.length) ? store.state.showThumbnails.filter(
            image => image.type === 'background',
        ) : '';

        if(bgImg !== '') {
            expect(dummyWrapper.vm.backgroundImage).toBe(bgImg[0].resolutions.original.url);
        } else {
            expect(dummyWrapper.vm.backgroundImage).toBe(bgImg);
        }
    });

    it('should validate the background image value', () => {
        const store = new Vuex.Store({
            state: {
                shows: showsData,
                genres: [],
                showDetails: showDetails,
                showThumbnails: [{
                    "id": 808,
                    "type": "poster",
                    "main": true,
                    "resolutions": {
                      "original": {
                        "url": "https://static.tvmaze.com/uploads/images/original_untouched/0/808.jpg"
                      },
                      "medium": {
                        "url": "https://static.tvmaze.com/uploads/images/medium_portrait/0/808.jpg"
                      }
                    }
                  }],
                showSearchbar: false
            },
            actions: {
                getShowDetails: jest.fn(),
                getShowThumbnails: jest.fn()
            }
        });
        const dummyWrapper = shallowMount(ShowDetails, {
            localVue,
            router,
            store,
            vuetify
        });
        let bgImg = (store.state.showThumbnails.length) ? store.state.showThumbnails.filter(
            image => image.type === 'background',
        ) : '';

        if(bgImg.length) {
            expect(dummyWrapper.vm.backgroundImage).toBe(bgImg[0].resolutions.original.url);
        } else {
            expect(dummyWrapper.vm.backgroundImage).toBe('');
        }
    });

    it('should validate the poster image value', () => {
        const store = new Vuex.Store({
            state: {
                shows: showsData,
                genres: [],
                showDetails: showDetails,
                showThumbnails: [{
                    "id": 546133,
                    "type": "background",
                    "main": false,
                    "resolutions": {
                      "original": {
                        "url": "https://static.tvmaze.com/uploads/images/original_untouched/219/549687.jpg"
                      }
                    }
                  }],
                showSearchbar: false
            },
            actions: {
                getShowDetails: jest.fn(),
                getShowThumbnails: jest.fn()
            }
        });
        const dummyWrapper = shallowMount(ShowDetails, {
            localVue,
            router,
            store,
            vuetify
        });
        let psImg = (store.state.showThumbnails.length) ? store.state.showThumbnails.filter(
            image => image.type === 'poster',
        ) : '';

        if(psImg.length) {
            expect(dummyWrapper.vm.posterImage).toBe(psImg[0].resolutions.original.url);
        } else {
            expect(dummyWrapper.vm.posterImage).toBe('');
        }
    });

    it('poster image data should match with store', () => {
        const pstImg = store.state.showThumbnails.filter(
            image => image.type === 'poster',
        );
        expect(wrapper.vm.posterImage).toBe(pstImg[0].resolutions.original.url);
    });

    it('goback method should redirect to home', () => {
        wrapper.vm.goBack();
        expect(wrapper.vm.$router.history.pending.name).toBe('showsCatalog');
    });

});