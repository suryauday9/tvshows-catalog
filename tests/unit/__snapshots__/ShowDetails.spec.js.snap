// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetails component renders component correctly 1`] = `
<div
  class="show-details"
>
  <v-btn-stub
    activeclass=""
    class="back-button black white--text"
    icon="true"
    tag="button"
    type="button"
  >
    <v-icon-stub>
      mdi-keyboard-return
    </v-icon-stub>
  </v-btn-stub>
   
  <div
    class="ma-5"
    style="background-size: 100% 96%; background-image: url(https://static.tvmaze.com/uploads/images/original_untouched/219/549687.jpg);"
  >
    <v-container-stub
      tag="div"
    >
      <v-row-stub
        class="black--text py-md-12"
        tag="div"
      >
        <v-col-stub
          class="show-thumbnail-block"
          cols="12"
          md="5"
          tag="div"
        >
          <v-img-stub
            class="ml-auto elevation-6"
            contain="true"
            options="[object Object]"
            position="center center"
            src="https://static.tvmaze.com/uploads/images/original_untouched/0/808.jpg"
            transition="fade-transition"
            width="390"
          />
        </v-col-stub>
         
        <v-col-stub
          class="pa-4 yellow lighten-5"
          cols="12"
          md="7"
          tag="div"
        >
          <h1
            class="display-2"
          >
            
              Justified
            
          </h1>
           
          <div
            class="d-flex flex-column pt-2"
          >
            <p
              class="subtitle-1 ma-0"
            >
              <v-chip-stub
                active="true"
                activeclass=""
                class="white--text purple ma-2"
                closeicon="$delete"
                closelabel="$vuetify.close"
                filtericon="$complete"
                tag="span"
              >
                <v-icon-stub
                  left=""
                  small=""
                >
                  mdi-information
                </v-icon-stub>
                
                    English
                  
              </v-chip-stub>
            </p>
             
            <p
              class="subtitle-1"
            >
              <v-chip-stub
                active="true"
                activeclass=""
                class="white--text teal ma-2"
                closeicon="$delete"
                closelabel="$vuetify.close"
                filtericon="$complete"
                tag="span"
              >
                <v-icon-stub
                  left=""
                  small=""
                >
                  mdi-information
                </v-icon-stub>
                
                    Drama
                
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="white--text teal ma-2"
                closeicon="$delete"
                closelabel="$vuetify.close"
                filtericon="$complete"
                tag="span"
              >
                <v-icon-stub
                  left=""
                  small=""
                >
                  mdi-information
                </v-icon-stub>
                
                    Crime
                
              </v-chip-stub>
              <v-chip-stub
                active="true"
                activeclass=""
                class="white--text teal ma-2"
                closeicon="$delete"
                closelabel="$vuetify.close"
                filtericon="$complete"
                tag="span"
              >
                <v-icon-stub
                  left=""
                  small=""
                >
                  mdi-information
                </v-icon-stub>
                
                    Western
                
              </v-chip-stub>
            </p>
             
            <p
              class="headline d-flex align-center"
            >
              
                8.9
                
              <v-icon-stub
                color="red"
              >
                mdi-star
              </v-icon-stub>
            </p>
             
            <div
              class="text-left"
            >
              <p>
                Deputy U.S. Marshal Raylan Givens is something of a 19th century-style officer in modern times, whose unconventional enforcement of justice makes him a target with criminals and a problem for his superiors in the U.S. Marshals Service. As a result of an eager but "justified" shooting of a Miami fugitive while on assignment in Florida, Givens is reassigned to Lexington, Kentucky, in a district that includes his hometown of Harlan, a rural mining town in the east of the state.
              </p>
            </div>
          </div>
        </v-col-stub>
      </v-row-stub>
    </v-container-stub>
  </div>
</div>
`;
